doc/crypto/api/ops/key-derivation.rst

```diff
@@ -847,19 +847,27 @@ Key derivation functions
 
     .. param:: const psa_key_attributes_t * attributes
         The attributes for the new key.
-        This function uses the attributes as follows:
 
-        *   The key type is required. It cannot be an asymmetric public key.
-        *   The key size is required. It must be a valid size for the key type.
-        *   The key permitted-algorithm policy is required for keys that will be used for a cryptographic operation, see :secref:`permitted-algorithms`.
+        The following attributes are required for all keys:
+
+        *   The key type. It must not be an asymmetric public key.
+        *   The key size. It must be a valid size for the key type.
+
+        The following attributes must be set for keys used in cryptographic operations:
+
+        *   The key permitted-algorithm policy, see :secref:`permitted-algorithms`.
 
             If the key type to be created is `PSA_KEY_TYPE_PASSWORD_HASH`, then the permitted-algorithm policy must be the same as the current operation's algorithm.
```

Comment:
Preexisting: technically, a more restrictive policy would also be ok. We currently have no wildcards that would be applicable here, but 0 should be allowed (even if I don't see a concrete use for it).

Reply:
The preceding text "for keys used in a cryptographic operation" would eliminate the zero case - as all operation functions demand one usage or another to be permitted. I think that the only thing one can do with a key with no usage flags is to export the public key, or destroy it (I haven't checked that, so might be missing something). Not sure it is worth documenting a technically valid, but entirely useless scenario for the specification?

I am inclined to leave this as is, and worry about relaxing the constraint when we have a relevant wildcard policy that would make the more complex-to-describe behavior valuable.

Reply:
I note now that this is not about `0` usage flags, but for a default no-permitted-algorithm policy. Keys like this can still be exported (if they have `PSA_KEY_USAGE_EXPORT`), and so can act as a secure storage for secrets.

There is also a test use case for simplified testing of key creation functionality where the construction of the key is tested by examining the resulting value, but the test suite can simply request no specific permitted algorithm.

I could reword this as a double negative, which perhaps implies that a 'no algorithm' policy is valid (but is not explicit):

> If the key type to be created is `PSA_KEY_TYPE_PASSWORD_HASH`, then the attributes must not specify a permitted-algorithm policy that is not the algorithm used to compute the password hash.

But that does not seem clear. If we wish to allow 'no permitted algorithm', then I think it would be better to explicitly include it in a positive specification rule:

> If the key type to be created is `PSA_KEY_TYPE_PASSWORD_HASH`, then the permitted-algorithm policy must either be the same as the current operation's algorithm, or `PSA_ALG_NONE`.

Is " the algorithm used to compute the password hash" as clear as "the current operation's algorithm"? - the former provides the rationale for the rule, but the implementor has to deduce that this is the operation algorithm; the latter is precise but the justification requires a deductive step.